File: desktop/renderer/src/settings/model-selection.ts
/**
 * Copyright © 2025-2026 Wenze Wei. All Rights Reserved.
 *
 * This file is part of Encre.
 * The Encre project belongs to the Dunimd Team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * DISCLAIMER: Users must comply with applicable AI regulations.
 * Non-compliance may result in service termination or legal liability.
 */

/**
 * Reusable "target model selection" widget (frontend half).
 *
 * A target-model selection is a list of model ids chosen to restrict which
 * configured models may serve a given feature (e.g. a gateway adapter's
 * allowed models in Settings -> Gateway).  This module is the UI + value
 * handling counterpart of ``encre.model_selection`` on the backend:
 *
 * * :func:`renderTargetModelSelection` -- build the selection card HTML
 *   (a toggle switch that reveals a multi-select list of enabled models).
 * * :func:`bindTargetModelSelection` -- wire up the toggle + click handlers
 *   after the HTML is inserted into the DOM.
 * * :func:`readTargetModelSelection` -- collect the currently selected
 *   model ids from the DOM.
 * * :func:`parseTargetModelSelection` / :func:`serializeTargetModelSelection`
 *   -- convert between the JSON string stored in settings (``adapter_<id>_models``)
 *   and the ``string[]`` in memory.
 *
 * The module is framework-agnostic and has no dependency on the settings
 * page: pass in the configured models, the initially selected ids and a
 * translation function, and it renders / binds / reads itself.
 */

import type { ModelConfigMeta } from "../core/types.js";

export interface TargetModelSelectionOptions {
  /** Unique id suffix for DOM ids (e.g. a gateway adapter id). */
  id: string;
  /** All configured models; enabled ones are shown as selectable items. */
  models: ModelConfigMeta[];
  /** Model ids selected when the widget renders. */
  selected: string[];
  /**
   * Explicitly force the enable toggle on/off.  When omitted the toggle
   * state is derived from ``selected`` (on when at least one model is
   * selected).  Use this to persist an "enabled but nothing picked yet"
   * state (e.g. per-workspace model restriction).
   */
  checked?: boolean;
  /** i18n lookup; defaults to the settings-page keys. */
  t: (key: string) => string;
  /** Optional i18n keys overriding the defaults. */
  titleKey?: string;
  hintKey?: string;
  emptyKey?: string;
}

function escapeHtml(str: string): string {
  return String(str)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

/** Parse a stored selection (JSON string) into a string[]; never throws. */
export function parseTargetModelSelection(raw: string | undefined | null): string[] {
  if (!raw) return [];
  try {
    const parsed = JSON.parse(raw);
    return Array.isArray(parsed) ? parsed.map(String) : [];
  } catch {
    return [];
  }
}

/** Serialize a selection into the JSON string persisted in settings. */
export function serializeTargetModelSelection(ids: string[]): string {
  return JSON.stringify(ids);
}

/** Build the selection card HTML. Insert the result, then call bindTargetModelSelection. */
export function renderTargetModelSelection(opts: TargetModelSelectionOptions): string {
  const { id } = opts;
  const t = opts.t;
  const titleKey = opts.titleKey ?? "settings.adapterTargetModel";
  const hintKey = opts.hintKey ?? "settings.adapterTargetModelHint";
  const emptyKey = opts.emptyKey ?? "settings.noModelConfigured";

  const title = t(titleKey);
  const hint = t(hintKey);
  const enabledModels = (opts.models ?? []).filter(m => m.enabled !== false);
  const selected = opts.selected ?? [];
  const on = opts.checked !== undefined ? !!opts.checked : selected.length > 0;

  const listHtml = enabledModels.length === 0
    ? `<span style="color:var(--text-muted);font-size:13px">${t(emptyKey)}</span>`
    : enabledModels.map(m => {
        const isSel = selected.includes(m.model_id);
        return `<div class="auto-push-gw-item${isSel ? " selected" : ""}" data-model-id="${escapeHtml(m.model_id)}">
              <span class="auto-push-gw-name">${escapeHtml(m.name || m.model_id)}</span>
              <span class="auto-push-gw-dot" title="${escapeHtml(m.model_id)}"></span>
              <span class="auto-push-gw-check"></span>
            </div>`;
      }).join("");

  return `
      <div class="settings-card model-selection-card" style="margin-top:12px;margin-bottom:0;overflow:hidden">
        <div class="settings-item-row">
          <div class="settings-item-info">
            <div class="settings-item-title">
              <span>${title}</span>
            </div>
            <div class="settings-item-desc">${hint}</div>
          </div>
          <div class="settings-item-control">
            <label class="toggle-switch" title="${title}">
              <input type="checkbox" id="adapter-model-toggle-${id}" ${on ? "checked" : ""} />
              <span class="toggle-slider"></span>
            </label>
          </div>
        </div>
        <div id="adapter-models-row-${id}" style="${on ? "" : "display:none"}">
          <div class="auto-push-gateways">
            <div class="auto-push-gateways-label">${title}</div>
            <div id="dlg-models-${id}" class="model-selection-list">${listHtml}</div>
          </div>
        </div>
      </div>`;
}

/** Bind the toggle switch and model item clicks inside the dialog. */
export function bindTargetModelSelection(
  root: HTMLElement,
  id: string,
  onChange?: (selected: string[]) => void,
): void {
  const container = root.querySelector<HTMLElement>(`#dlg-models-${id}`);
  const row = root.querySelector<HTMLElement>(`#adapter-models-row-${id}`);
  const toggle = root.querySelector<HTMLInputElement>(`#adapter-model-toggle-${id}`);
  const emit = (): void => {
    if (onChange) onChange(readTargetModelSelection(root, id));
  };
  if (toggle && row) {
    toggle.addEventListener("change", () => {
      row.style.display = toggle.checked ? "" : "none";
      emit();
    });
  }
  if (!container) return;
  container.querySelectorAll(".auto-push-gw-item").forEach(item => {
    item.addEventListener("click", (e) => {
      e.stopPropagation();
      item.classList.toggle("selected", !item.classList.contains("selected"));
      emit();
    });
  });
}

/** Read the currently selected model ids from the DOM. */
export function readTargetModelSelection(root: HTMLElement, id: string): string[] {
  const container = root.querySelector<HTMLElement>(`#dlg-models-${id}`);
  if (!container) return [];
  return Array.from(container.querySelectorAll<HTMLElement>(".auto-push-gw-item.selected"))
    .map(item => item.getAttribute("data-model-id")!)
    .filter(Boolean);
}

/** Read whether the enable toggle is currently switched on. */
export function readTargetModelSelectionEnabled(root: HTMLElement, id: string): boolean {
  const toggle = root.querySelector<HTMLInputElement>(`#adapter-model-toggle-${id}`);
  return toggle ? toggle.checked : false;
}

File: core/encre/model_selection.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Copyright © 2025-2026 Wenze Wei. All Rights Reserved.
#
# This file is part of Encre.
# The Encre project belongs to the Dunimd Team.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# DISCLAIMER: Users must comply with applicable AI regulations.
# Non-compliance may result in service termination or legal liability.

"""Reusable "target model selection" logic.

The target-model-selection feature lets a consumer (e.g. a gateway adapter
configured via Settings -> Gateway -> Target Models) restrict which
configured models may serve its runs.  This module centralises the
transport-agnostic pieces shared by every consumer:

* :func:`parse_target_model_ids` -- turn a stored selection (a JSON string
  or a raw list, the shape persisted via the flat ``adapter_<id>_models``
  config key) into a list of model ids, or ``None`` when unrestricted.
* :func:`target_models_from_config_dict` -- read a selection straight out
  of a stored per-consumer config dict.
* :func:`apply_target_models` -- restrict a config copy to the given target
  models and pin it to the first eligible candidate so a run starts there.

Both the gateway runner and any future consumer (automation, relay, ...)
can reuse these functions without knowing each other's storage layout.
"""

import json
from typing import Any

from encre.config import EncreConfig, ModelConfig


def parse_target_model_ids(raw: Any) -> list[str] | None:
    """Parse a stored target-model selection into a list of model ids.

    Accepts either a JSON-encoded string (the shape written by the
    frontend) or a raw list.  Returns ``None`` when the value is empty,
    malformed or not a non-empty list -- i.e. the consumer is unrestricted
    and should fall back to its normal model resolution.
    """
    if raw is None or raw == "":
        return None
    if isinstance(raw, str):
        try:
            raw = json.loads(raw)
        except Exception:
            return None
    if not isinstance(raw, list) or not raw:
        return None
    return [str(i) for i in raw]


def target_models_from_config_dict(stored: dict[str, Any]) -> list[str] | None:
    """Read the target model selection out of a stored config dict.

    The selection is stored under the ``models`` field (``models`` is also
    the field name the frontend persists as ``adapter_<id>_models``); the
    ``model_ids`` alias is accepted for legacy configs.
    """
    raw = stored.get("models") if stored.get("models") is not None else stored.get("model_ids")
    return parse_target_model_ids(raw)


def apply_target_models(
    cfg: EncreConfig,
    target_model_ids: list[str] | None,
) -> ModelConfig | None:
    """Restrict *cfg* to the given target models and pin it to the first
    eligible candidate.

    Sets ``cfg.target_model_ids`` so the agent's runtime model-failure
    fallback only considers those models, then pins ``cfg`` onto the first
    eligible candidate (target models first, then indicator, then a random
    enabled model).  Returns the pinned model, or ``None`` when no target
    model was requested or no candidate is eligible.
    """
    if not target_model_ids:
        return None
    cfg.target_model_ids = list(target_model_ids)
    candidates = cfg.resolve_model_candidates()
    if not candidates:
        return None
    first = candidates[0]
    cfg.model = first.model_id
    cfg.backend_type = first.backend_type
    cfg.api_key = first.api_key
    cfg.base_url = first.base_url
    cfg.max_tokens = first.max_tokens
    return first
